pkg/config/signer.go

```diff
@@ -0,0 +1,244 @@
+	// They are precomputed for efficiency reasons.
+	curveHalfOrders = map[elliptic.Curve]*big.Int{
+		elliptic.P224(): new(big.Int).Rsh(elliptic.P224().Params().N, 1),
+		elliptic.P256(): new(big.Int).Rsh(elliptic.P256().Params().N, 1),
+		elliptic.P384(): new(big.Int).Rsh(elliptic.P384().Params().N, 1),
+		elliptic.P521(): new(big.Int).Rsh(elliptic.P521().Params().N, 1),
+	}
+)
+
+// Signer holds the information necessary to sign
+// a configuration update.
+type Signer struct {
+	cert       *x509.Certificate
+	mspID      string
+	privateKey *ecdsa.PrivateKey
+	publicKey  *ecdsa.PublicKey
+}
+
+type ecdsaSignature struct {
+	R, S *big.Int
+}
+
+// NewSigner creates a new signer for configuration updates. The publicCert
+// and privateKey should be pem encoded.
+func NewSigner(publicCert []byte, privateCert []byte, mspID string) (*Signer, error) {
+	if mspID == "" {
+		return nil, errors.New("failed to create new signer, mspID can not be empty")
+	}
+
+	cert, err := getCertFromPem(publicCert)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get cert from pem: %v", err)
+	}
+
+	publicKey, err := ecdsaPublicKeyImport(cert)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ECDSA public key: %v", err)
+	}
+
+	pkBlock, _ := pem.Decode(privateCert)
+	if pkBlock == nil {
+		return nil, errors.New("failed to decode private key from pem")
+	}
+
+	privateKey, err := ecdsaPrivateKeyImport(pkBlock.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ECDSA private key: %v", err)
+	}
+
+	signer := &Signer{
+		cert:       cert,
+		mspID:      mspID,
+		privateKey: privateKey,
+		publicKey:  publicKey,
+	}
+	return signer, nil
+}
+
+// Serialize returns a byte array representation of this identity.
+func (s *Signer) Serialize() ([]byte, error) {
+	pb := &pem.Block{Bytes: s.cert.Raw, Type: "CERTIFICATE"}
+	pemBytes := pem.EncodeToMemory(pb)
+	if pemBytes == nil {
+		return nil, errors.New("failed to encode pem block")
+	}
+
+	// serialize identities by prepending the MSPID and appending
+	// the ASN.1 DER content of the cert
+	sID := &msp.SerializedIdentity{Mspid: s.mspID, IdBytes: pemBytes}
+	idBytes := MarshalOrPanic(sID)
+
+	return idBytes, nil
+}
+
+// Public returns the public key of the signer.
+func (s *Signer) Public() crypto.PublicKey {
+	return s.publicKey
+}
+
+// Cert returns the certificate of the signer.
+func (s *Signer) Cert() *x509.Certificate {
+	return s.cert
+}
+
+// MSPId returns the MSP ID of the signer.
+func (s *Signer) MSPId() string {
+	return s.mspID
+}
+
+// Sign performs ECDSA sign with signer's private key on given digest.
+func (s *Signer) Sign(reader io.Reader, digest []byte) (signature []byte, err error) {
+	if reader == nil {
+		return nil, errors.New("failed to sign, reader can not be nil")
+	}
+
+	rr, ss, err := ecdsa.Sign(reader, s.privateKey, digest)
+	if err != nil {
+		return nil, err
+	}
+
+	ss, _, err = toLowS(&s.privateKey.PublicKey, ss)
```

Comment:
If this is to be in _pkg_ then let's at least _document_ why we're doing this.

Reply:
Document the "toLowS" part? We could use some input from some crypto experts on what some of these operations actually represent. As is, we pulled most of the crypto code here from the fabric bccsp/msp packages.

Reply:
Ah, found another place in fabric where we have a "toLowS" function. And it's documented! https://github.com/hyperledger/fabric/blob/master/internal/cryptogen/csp/csp.go#L137-L143 We'll incorporate that info here.

Reply:
Done.